mailpoet/lib/Automation/Engine/Validation/WorkflowRules/UnknownStepRule.php

```diff
@@ -25,36 +22,31 @@ class WorkflowStepsValidator {
 
   public function __construct(
     Registry $registry,
-    Validator $validator,
     WorkflowStorage $workflowStorage
   ) {
-    $this->validator = $validator;
     $this->registry = $registry;
     $this->workflowStorage = $workflowStorage;
   }
 
-  public function validateSteps(Workflow $workflow): void {
-    foreach ($workflow->getSteps() as $step) {
-      $this->validateStep($workflow, $step);
-    }
+  public function initialize(Workflow $workflow): void {
+    $this->cachedExistingWorkflow = false;
```

Comment:
Just reading what `cachedExistingWorkflow` does. Should we do caching on that level or thinking about this at a deeper level like

```
class WorkflowStorageCache implements WorkflowStorage {

	private $dbStorage;
	
	public function __construct(WorkflowStorageDb $dbStorage) {
		$this->dbStorage = $dbStorage;
	}
	
	public function getById(int $id) {
		if (cached) {
			return cached;
		}
		
		cached = $this->dbStorage->getbyId($id);
		do_cache
		return cached;
	}
	
	public function update(Workflow) {
		delete_cache
		$this->dbStorage->update(Workflow);
	}
}
```

Reply:
Good idea, we really could try to make sure some of the main entities are fetched only once per request. I added this to the P2 ideas draft.